lib/Parse/ParseExpr.cpp

```diff
@@ -3122,19 +3138,106 @@ ParserStatus Parser::parseExprList(tok leftTok, tok rightTok,
     return status;
 
   // Parse the closure.
-  ParserResult<Expr> closure =
-    parseTrailingClosure(SourceRange(leftLoc, rightLoc));
-  status |= closure;
-  if (closure.isNull())
-    return status;
+  status |=
+      parseTrailingClosures(isExprBasic, SourceRange(leftLoc, rightLoc),
+                            trailingLBrace, trailingRBrace, trailingClosures);
+  return status;
+}
 
-  // Record the trailing closure.
-  trailingClosure = closure.get();
+/// Check whether upcoming trailing closure actually represents a block of
+/// labeled trailing closures e.g.
+/// foo() {
+///  <label-1>: { ... }
+///  ...
+/// }
+static bool isBlockOfMultipleTrailingClosures(bool isExprBasic, Parser &P) {
+  assert(isValidTrailingClosure(isExprBasic, P));
+
+  // If closure doesn't start from a label there is no reason
+  // to look any farther.
+  {
+    const auto &nextTok = P.peekToken();
+    if (!nextTok.canBeArgumentLabel() || nextTok.is(tok::kw__) ||
+        nextTok.getText()[0] == '$')
+      return false;
+  }
 
-  return status;
+  Parser::BacktrackingScope backtrack(P);
+  P.consumeToken(tok::l_brace);
+  P.consumeToken(); // Consume Label text.
+
+  if (!P.consumeIf(tok::colon)) // Consume `:`
+    return false;
+
+  // If the next token after the label is not a left brace
+  // that means what follows is not a valid closure, which
+  // in turn means it can't be a labeled block.
+  if (!P.Tok.is(tok::l_brace))
+    return false;
+
+  return isValidTrailingClosure(isExprBasic, P);
 }
 
-ParserResult<Expr> Parser::parseTrailingClosure(SourceRange calleeRange) {
+ParserStatus Parser::parseMultipleTrailingClosures(
+    SourceLoc &LBrace, SourceLoc &RBrace,
+    SmallVectorImpl<TrailingClosure> &closures) {
+  SyntaxParsingContext ClauseCtx(SyntaxContext,
+                                 SyntaxKind::MultipleTrailingClosureClause);
+
+  // Consume '{' of the trailing closure
+  LBrace = consumeToken(tok::l_brace);
+
+  ParserStatus Status;
+  // There could N labeled closures depending on the number of arguments.
+  do {
+    if (!(Tok.canBeArgumentLabel() && peekToken().is(tok::colon))) {
+      Status.setIsParseError();
+      diagnose(Tok.getLoc(),
+               diag::expected_argument_label_followed_by_closure_literal);
+      skipUntilDeclStmtRBrace(tok::r_brace);
+      break;
+    }
+
+    SyntaxParsingContext ElementCtx(SyntaxContext,
+                                    SyntaxKind::MultipleTrailingClosureElement);
+    Identifier label;
+    auto labelLoc = consumeArgumentLabel(label);
+    if (!labelLoc.isValid())
+      return makeParserError();
+
+    consumeToken(); // Consume ':'
+
+    auto closure = parseExprClosure();
```

Comment:
`Tok.is(tok::l_brace)` check is needed before calling this. `parseExprClosure()` assert it.

Reply:
Or just `parseExpr()` and diagnose if the result is not `ClosureExpr`/`CaptureListExpr`